lgc/state/PalMetadata.cpp

```diff
@@ -99,3 +136,126 @@ void PalMetadata::record(Module *module) {
+    else
+      m_userDataRegMapping[ShaderStageTessEval] = m_userDataRegMapping[ShaderStageGeometry];
+  } else if (m_pipelineState->getNggControl()->enableNgg) {
+    m_userDataRegMapping[ShaderStageTessEval] = mmSPI_SHADER_USER_DATA_GS_0;
+  } else {
+    m_userDataRegMapping[ShaderStageTessEval] = mmSPI_SHADER_USER_DATA_VS_0;
+  }
+
+  m_userDataRegMapping[ShaderStageTessControl] = mmSPI_SHADER_USER_DATA_HS_0;
+  if (m_pipelineState->hasShaderStage(ShaderStageTessControl)) {
+    if (m_pipelineState->getTargetInfo().getGfxIpVersion().major < 9)
+      m_userDataRegMapping[ShaderStageVertex] = mmSPI_SHADER_USER_DATA_LS_0;
+    else
+      m_userDataRegMapping[ShaderStageVertex] = m_userDataRegMapping[ShaderStageTessControl];
+  } else if (m_pipelineState->hasShaderStage(ShaderStageGeometry)) {
+    if (m_pipelineState->getTargetInfo().getGfxIpVersion().major < 9)
+      m_userDataRegMapping[ShaderStageVertex] = mmSPI_SHADER_USER_DATA_ES_0;
+    else
+      m_userDataRegMapping[ShaderStageVertex] = m_userDataRegMapping[ShaderStageGeometry];
+  } else if (m_pipelineState->getNggControl()->enableNgg) {
+    m_userDataRegMapping[ShaderStageVertex] = mmSPI_SHADER_USER_DATA_GS_0;
+  } else {
+    m_userDataRegMapping[ShaderStageVertex] = mmSPI_SHADER_USER_DATA_VS_0;
+  }
+
+  return m_userDataRegMapping[stage];
+}
+
+// =====================================================================================================================
+// Set the PAL metadata SPI register for a number of consecutive user data entries
+//
+// @param stage : ShaderStage
+// @param userDataIndex : User data index 0-15 or 0-31 depending on HW and shader stage
+// @param userDataValue : Value to store in that entry, one of:
+//                        - a 0-based integer for the root user data dword offset
+//                        - DescRelocMagic+set_number for an unlinked descriptor set number
+//                        - one of the UserDataMapping values, e.g. UserDataMapping::GlobalTable
+// @param dwordCount : Number of user data entries to set
+void PalMetadata::setUserDataEntry(ShaderStage stage, unsigned userDataIndex, unsigned userDataValue,
+                                   unsigned dwordCount) {
+  // If this is the spill table pointer, adjust userDataLimit accordingly.
+  if (userDataValue == static_cast<unsigned>(Util::Abi::UserDataMapping::SpillTable))
+    setUserDataLimitToMax();
+
+  // Get the start register number of SPI user data registers for this shader stage.
+  unsigned userDataReg = getUserDataReg0(stage);
+
+  // Assert that the supplied user data index is not too big.
+  assert(userDataIndex + dwordCount <= 32 &&
+         (m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9 || stage != ShaderStageCompute ||
+          userDataIndex + dwordCount <= 16) &&
+         "Out of range user data index");
+
+  // Update userDataLimit if needed.
+  if (userDataValue < 0x10000 && userDataValue + dwordCount > m_userDataLimit->getUInt())
```

Comment:
What's the meaning of 0x10000?

Reply:
Good point. :-) I'll use a named constant and improve the comment.